src/hypothesis/searchstrategy/regex.py

```diff
@@ -0,0 +1,150 @@
+        elif category == 'category_not_digit':
+            return string.ascii_letters + string.punctuation
+        elif category == 'category_space':
+            return string.whitespace
+        elif category == 'category_not_space':
+            return string.printable.strip()
+        elif category == 'category_word':
+            return string.ascii_letters + string.digits + '_'
+        elif category == 'category_not_word':
+            return ''.join(set(string.printable)
+                           .difference(string.ascii_letters +
+                                       string.digits + '_'))
+        else:
+            raise NotImplementedError
+
+    def _handle_character_sets(self, state):
+        opcode, value = state
+        if opcode == 'range':
+            return [hunichr(val) for val in hrange(value[0], value[1] + 1)]
+        elif opcode == 'literal':
+            return [hunichr(value)]
+        elif opcode == 'category':
+            return self._categories(value)
+        else:
+            raise NotImplementedError
+
+    def _handle_state(self, state):
+        opcode, value = state
+        if opcode == 'literal':
+            return strats.just(hunichr(value))
+        elif opcode == 'not_literal':
+            return strats.characters(blacklist_characters=hunichr(value))
+        elif opcode == 'at':
+            return strats.just('')
+        elif opcode == 'in':
+            if value[0][0] == 'negate':
+                candidates = []
+                for v in value[1:]:
+                    candidates.extend(chain(*(self._handle_character_sets(v))))
+                return strats.characters(blacklist_characters=candidates)
+            else:
+                candidates = []
+                for v in value:
+                    candidates.extend(chain(*(self._handle_character_sets(v))))
+                return strats.sampled_from(candidates)
+        elif opcode == 'any':
+            return strats.characters()
+        elif opcode == 'branch':
+            branches = []
+            for val in value[1]:
+                branch = [self._handle_state(v) for v in val]
+                branches.append(strategy_concat(branch))
+            return strats.one_of(branches)
+        elif opcode == 'subpattern':
+            parts = []
+            for part in value[1]:
+                parts.append(self._handle_state(part))
+            result = strategy_concat(parts)
+            if value[0]:
+                self.cache[value[0]] = result
+                result = strats.shared(result, key=value[0])
+            return result
+        elif opcode == 'assert':
+            result = []
+            for part in value[1]:
+                result.append(self._handle_state(part))
+            return strategy_concat(result)
+        elif opcode == 'assert_not':
+            return strats.just('')
+        elif opcode == 'groupref':
+            return strats.shared(self.cache[value], key=value)
```

Comment:
Yeah, I don't think shared is a good way of doing this. it would cause a bunch of problems. e.g. if you had `lists(regex(r"<([a-z]+)>.*</\1>"))` then every regex in the list would have to have the same value for `\1`